src/python_testing/TC_DISHM_2_1.py

```diff
@@ -0,0 +1,159 @@
+
+        asserts.assert_true(self.check_pics("DISHM.S.A0000"), "DISHM.S.A0000 must be supported")
+        asserts.assert_true(self.check_pics("DISHM.S.A0001"), "DISHM.S.A0001 must be supported")
+        asserts.assert_true(self.check_pics("DISHM.S.C00.Rsp"), "DISHM.S.C00.Rsp must be supported")
+        asserts.assert_true(self.check_pics("DISHM.S.C01.Tx"), "DISHM.S.C01.Tx must be supported")
+
+        attributes = Clusters.DishwasherMode.Attributes
+
+        self.print_step(1, "Commissioning, already done")
+
+        self.print_step(2, "Read SupportedModes attribute")
+        supported_modes = await self.read_mod_attribute_expect_success(endpoint=self.endpoint, attribute=attributes.SupportedModes)
+
+        logging.info("SupportedModes: %s" % (supported_modes))
+
+        asserts.assert_greater_equal(len(supported_modes), 2, "SupportedModes must have at least two entries!")
+
+        supported_modes_dut = [m.mode for m in supported_modes]
+
+        self.print_step(3, "Read CurrentMode attribute")
+
+        old_current_mode_dut = await self.read_mod_attribute_expect_success(endpoint=self.endpoint, attribute=attributes.CurrentMode)
+
+        logging.info("CurrentMode: %s" % (old_current_mode_dut))
+
+        # pick a value that's not in the list of supported modes
+        invalid_mode_th = max(supported_modes_dut) + 1
+
+        from enum import Enum
+
+        class CommonCodes(Enum):
+            SUCCESS = 0x00
+            UNSUPPORTED_MODE = 0x01
+            GENERIC_FAILURE = 0x02
+
+        DishwasherModeCodes = [code.value for code in Clusters.DishwasherMode.Enums.StatusCode
+                               if code is not Clusters.DishwasherMode.Enums.StatusCode.kUnknownEnumValue]
+
+        self.print_step(4, "Send ChangeToMode command with NewMode set to %d" % (old_current_mode_dut))
+
+        ret = await self.send_change_to_mode_cmd(newMode=old_current_mode_dut)
+        asserts.assert_true(ret.status == CommonCodes.SUCCESS.value, "Changing the mode to the current mode should be a no-op")
+
+        if self.check_pics("DISHM.S.M.CAN_TEST_MODE_FAILURE"):
+            self.print_step(5, "Manually put the device in a state from which it will FAIL to transition to mode %d" % (self.modefail))
+            input("Press Enter when done.\n")
+
+            self.print_step(6, "Read CurrentMode attribute")
+            old_current_mode_dut = await self.read_mod_attribute_expect_success(endpoint=self.endpoint, attribute=attributes.CurrentMode)
+
+            logging.info("CurrentMode: %s" % (old_current_mode_dut))
+
+            self.print_step(7, "Send ChangeToMode command with NewMode set to %d" % (self.modefail))
+
+            ret = await self.send_change_to_mode_cmd(newMode=self.modefail)
+            st = ret.status
+            is_mfg_code = st in range(0x80, 0xC0)
+            is_err_code = (st == CommonCodes.GENERIC_FAILURE.value) or (st in DishwasherModeCodes) or is_mfg_code
```

Comment:
OK, so now this allows the errors in DishwasherModeCodes... but again: the test plan does not allow that.  Is there an issue tracking this being allowed?

And again, why is InvalidInMode not allowed as a failure here?  Is there a test plan issue tracking that?

All of which I already asked back in August.....

I totally believe that TC_RVCRUNM_2_1.py and the test plan for it has the same issues, by the way.  That just means we need to get them filed and fixed.

Reply:
There is currently no invalid mode check in step 7. Invalid mode isn't tested until step 13.

Reply:
My point is: why can't the device respond with InvalidInMode in step 7?